### ISSUE TITLE
Advanced Q problem

### ISSUE THREAD

**Issue by Ankhbayar** [hmarr/mongoengine#62]
What happened in Q **and** , **or** method. 
Something is wrong. 

```
import mongoengine as mongo
from mongoengine.queryset import Q

# Test Q and Filter combination

mongo.connect("test_q")

class User(mongo.Document):
    name = mongo.StringField()

User.drop_collection()


class Post(mongo.Document):
    title = mongo.StringField()
    rank = mongo.IntField()
    tags = mongo.ListField(mongo.StringField())
    created_by = mongo.ReferenceField(User)
    contrib_users = mongo.ListField(mongo.ReferenceField(User))
    donated_users = mongo.ListField(mongo.ReferenceField(User))

Post.drop_collection()


# Prepare Data
bold = User.objects.create(name = 'bold')
bayar = User.objects.create(name = 'bayar')
naran = User.objects.create(name = 'naran')

tulga = User.objects.create(name = 'tulga')
tsetsen = User.objects.create(name = 'tsetsen')

# Bold in all fields
Post.objects.create( 
    title = 'test1', 
    rank = 1, 
    tags = ['mongoengine', 'mongo', 'pymongo'], 
    created_by = bold,
    contrib_users = [bold, bayar],
    donated_users = [tulga, tsetsen, bayar, bold ]
    )

# Bold only contrib_users
Post.objects.create( 
    title = 'test2', 
    rank = 2, 
    tags = ['mongo'], 
    created_by = bayar,
    contrib_users = [bold, bayar],
    donated_users = [tulga, tsetsen, bayar, naran ],
)

# Bold no related
Post.objects.create( 
    title = 'test3', 
    rank = 5, 
    tags = ['mongo','run', 'mongodb' ], 
    created_by = tsetsen,
    contrib_users = [ bayar, naran ],
    donated_users = [ tulga, tsetsen, bayar, naran ],
)


# Case 1 -----------------
bold_related_tags_mongo = Post.objects.filter(
                    Q(rank__gt = 0) &
                    (
                        Q(contrib_users = bold) |
                        Q(donated_users = bold) |
                        Q(created_by = bold)
                    ) &
                    Q(tags = 'mongo')
)

len(bold_related_tags_mongo) # 2 - OK

bold_related_tags_mongo = Post.objects.filter(
                    Q(rank__gt = 0) &
                    Q(tags = 'mongo') &
                    (
                        Q(contrib_users = bold) |
                        Q(donated_users = bold) |
                        Q(created_by = bold)
                    )
)

len(bold_related_tags_mongo) # 1 - ERROR

# Case 2 -----------------

bold_related_tags_mongo = Post.objects.filter(tags = 'mongo')

bold_related_tags_mongo = bold_related_tags_mongo.filter(
                    Q(rank__gt = 0) &
                    (
                        Q(contrib_users = bold) |
                        Q(donated_users = bold) |
                        Q(created_by = bold)
                    )
)

for post in bold_related_tags_mongo: # once loop - ERROR
    if post.title not in ['test1','test2',]:
        print "ERROR", post.title
    else:
        print post.title

len(bold_related_tags_mongo) # 1 - ERROR

# Case 3 -----------------

bold_related_tags_mongo = Post.objects.filter(tags = 'mongo')

len(bold_related_tags_mongo)

bold_related_tags_mongo = bold_related_tags_mongo.filter(
                    Q(rank__gt = 0) &
                    (
                        Q(contrib_users = bold) |
                        Q(donated_users = bold) |
                        Q(created_by = bold)
                    )
)

for post in bold_related_tags_mongo: # once loop - ERROR
    if post.title not in ['test1','test2',]:
        print "ERROR", post.title
    else:
        print post.title

len(bold_related_tags_mongo) # 3 - ERROR
```

**Comment by flosch** [comment]
Which version/branch do you use?

**Comment by Ankhbayar** [comment]
v0.4 Branch checkout.

**Comment by Ankhbayar** [comment]
sorry I forgot write version.
checkout from Branch v0.4 at 2010/07/28

**Comment by hmarr** [comment]
Delete stale cursor when query is filtered. Closed by 4742328b908e57b3d87e2c0752ed8d4b51763e55.

**Comment by Ankhbayar** [comment]
Great. Thanks Hmarr.
